src/ReplaySession.cc

```diff
@@ -298,9 +308,17 @@ Completion ReplaySession::enter_syscall(ReplayTask* t,
       // overwritten by the tracee. It could even be dynamically generated and
       // not generated yet.
       if (t->vm()->is_breakpoint_in_private_read_only_memory(
-              syscall_instruction) &&
-          t->vm()->add_breakpoint(syscall_instruction, BKPT_INTERNAL)) {
-        use_breakpoint_optimization = true;
+              syscall_instruction)) {
+        // Check that the breakpoint didn't disappear underneath our feet
+        if (syscall_bp_vm &&
+            t->vm()->get_breakpoint_type_at_addr(syscall_instruction) == BKPT_NONE)
+          syscall_bp_vm = nullptr;
```

Comment:
ReplayTimeline unapplies all breakpoints, but only reapplies those it knows about (which does not include the internal ones).